frame/pallet-xcm-transactor/src/chain_extension.rs

```diff
@@ -0,0 +1,281 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use crate::{Config, Pallet, QueryConfig};
+use frame_support::{traits::EnsureOrigin, DefaultNoBound};
+use frame_system::RawOrigin;
+use pallet_contracts::chain_extension::{
+    ChainExtension, Environment, Ext, InitState, Result as DispatchResult, RetVal, SysConfig,
+};
+use pallet_xcm::{Pallet as XcmPallet, WeightInfo};
+use parity_scale_codec::Encode;
+use sp_std::prelude::*;
+use xcm::prelude::*;
+pub use xcm_ce_primitives::{
+    Command::{self, *},
+    PreparedExecution, ValidateSendInput, ValidatedSend,
+    XcmCeError::{self, *},
+    XCM_EXTENSION_ID,
+};
+use xcm_executor::traits::WeightBounds;
+
+type RuntimeCallOf<T> = <T as SysConfig>::RuntimeCall;
+
+macro_rules! unwrap {
+    ($val:expr, $err:expr) => {
+        match $val {
+            Ok(inner) => inner,
+            Err(_) => return Ok(RetVal::Converging($err.into())),
+        }
+    };
+}
+
+#[derive(DefaultNoBound)]
+pub struct XCMExtension<T: Config> {
+    prepared_execute: Option<PreparedExecution<RuntimeCallOf<T>>>,
+    validated_send: Option<ValidatedSend>,
+}
+
+impl<T: Config> ChainExtension<T> for XCMExtension<T>
+where
+    <T as SysConfig>::AccountId: AsRef<[u8; 32]>,
+{
+    fn call<E>(&mut self, env: Environment<E, InitState>) -> DispatchResult<RetVal>
+    where
+        E: Ext<T = T>,
+    {
+        match unwrap!(env.func_id().try_into(), InvalidCommand) {
+            PrepareExecute => self.prepare_execute(env),
+            Execute => self.execute(env),
+            ValidateSend => self.validate_send(env),
+            Send => self.send(env),
+            NewQuery => self.new_query(env),
+            TakeResponse => self.take_response(env),
+            PalletAccountId => self.pallet_account_id(env),
+        }
+    }
+}
+
+impl<T: Config> XCMExtension<T> {
+    /// Returns the weight for given XCM and saves it (in CE, per-call scratch buffer) for
+    /// execution
+    fn prepare_execute<E: Ext<T = T>>(
+        &mut self,
+        env: Environment<E, InitState>,
+    ) -> DispatchResult<RetVal> {
+        let mut env = env.buf_in_buf_out();
+        // input parsing
+        let len = env.in_len();
+        let input: VersionedXcm<RuntimeCallOf<T>> = env.read_as_unbounded(len)?;
+
+        let mut xcm = unwrap!(input.try_into(), BadVersion);
+        // calculate the weight
+        let weight = unwrap!(T::Weigher::weight(&mut xcm), CannotWeigh);
+
+        // save the prepared xcm
+        self.prepared_execute = Some(PreparedExecution { xcm, weight });
+        // write the output to buffer
+        weight.using_encoded(|w| env.write(w, true, None))?;
+
+        Ok(RetVal::Converging(XcmCeError::Success.into()))
+    }
```

Comment:
What's the weight associated with this?

Can it be spammed until resources are exhausted?

Reply:
Nothing as of now, weights are overall a TODO still.
Best choice would be moving most of this logic to pallet and benchmark that for weights.